File: src/main/java/de/thyroff/imgsteg/ByteArrayReader.java
package de.thyroff.imgsteg;

import de.thyroff.imgsteg.utils.BitBuffer;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.util.logging.Level;
import java.util.logging.Logger;

public class ByteArrayReader {
    private static final Logger logger = Logger.getLogger(ByteArrayReader.class.getName());

    /**
     * Reads Byte Stream from the provided Image. Simply takes the lowest bits of RGB color channels without any other logic. Starting at 0 / 0.
     * @param image The image file into which the byte stream will be injected.
     * @throws IOException If an I/O error occurs while reading the image file.
     */
    public static byte[] readByteArrayFromImage(File image) throws IOException {
        logger.log(Level.FINER, "read byte stream from image");

        BufferedImage bufferedImage = ImageIO.read(image);

        final int width = bufferedImage.getWidth();

        int pixelIndex = 0;

        BitBuffer bitBuffer = new BitBuffer();

        //determine buffer length
        //read first 33 bit: length of buffer array
        for (int i = 0; i < 11; i++) {
            bitBuffer.add(bufferedImage, width, pixelIndex);
            pixelIndex++;
        }
        final int bufferSize = bitBuffer.removeInt(); //positions to be read - ie. size of the following buffer

        assert bitBuffer.size() == 1; //first bit of first byte already in the buffer

        //now read the rest of the bytes
        while (bitBuffer.size() < bufferSize * 8){
            bitBuffer.add(bufferedImage, width, pixelIndex);
            pixelIndex++;
        }
        logger.log(Level.FINER, "Byte stream read.");

        //convert bitBuffer to byte array
        return bitBuffer.removeByteArray(bufferSize);
    }
}

File: reactimgsteg/src/utils/ByteArrayReader.js
import { BitBuffer } from "./BitBuffer";
export class ByteArrayReader {

  /**
   * Reads a byte stream from the provided image data. It simply takes the lowest bits of RGB color channels.
   * Assumes the image is already loaded and rendered onto a canvas.
   * @param {ImageData} imageData The ImageData object from which the byte stream will be extracted.
   * @returns {Uint8Array} The extracted byte array.
   */
  static readByteArrayFromImage(imageData) {
    console.log("read byte stream from image");

    const width = imageData.width;
    let pixelIndex = 0;

    const bitBuffer = new BitBuffer();

    // Determine buffer length - read first 33 bits: length of buffer array
    // Assuming that the image is large enough to hold at least 33 bits
    for (let i = 0; i < 11; i++) {
      this.addPixelDataToBitBuffer(imageData, width, pixelIndex, bitBuffer);
      pixelIndex++;
    }

    const bufferSize = bitBuffer.removeInt(); // Positions to be read - i.e., size of the following buffer

    if (bitBuffer.size() !== 1) {
      throw new Error('Buffer size mismatch.');
    }

    // Now read the rest of the bytes
    while (bitBuffer.size() < bufferSize * 8) {
      this.addPixelDataToBitBuffer(imageData, width, pixelIndex, bitBuffer);
      pixelIndex++;
    }

    console.log("Byte stream read.");

    // Convert bitBuffer to byte array
    return bitBuffer.removeByteArray(bufferSize);
  }

  /**
   * Adds pixel data to the BitBuffer object.
   * @param {ImageData} imageData The image data.
   * @param {number} width The width of the image.
   * @param {number} pixelIndex The current pixel index.
   * @param {BitBuffer} bitBuffer The BitBuffer instance.
   */
  static addPixelDataToBitBuffer(imageData, width, pixelIndex, bitBuffer) {
    const x = pixelIndex % width;
    const y = Math.floor(pixelIndex / width);
    const i = (y * width + x) * 4; // pixel index in the imageData array

    const r = imageData.data[i];
    const g = imageData.data[i + 1];
    const b = imageData.data[i + 2];

    bitBuffer.add(r & 1); // Add the least significant bit of the red channel
    bitBuffer.add(g & 1); // Add the least significant bit of the green channel
    bitBuffer.add(b & 1); // Add the least significant bit of the blue channel
  }
}
